### PR TITLE
ARROW-16510: [R] Add bindings for GCS filesystem

### PR DESCRIPTION
**Pull request by nealrichardson**
This adds basic bindings for GcsFileSystem to R, turns it on in the macOS, Windows, and Linux packaging (same handling as ARROW_S3), and basic R tests.

Followups: 

- Bindings for FromImpersonatedServiceAccount (ARROW-16885)
- Set up testbench for fuller tests, like how we do with minio (ARROW-16879)
- GcsFileSystem::Make should return Result (ARROW-16884)
- Explore auth integration/compatibility with `gargle`, `googleAuthR`, etc.: can we pick up the same credentials they use (ARROW-16880)
- macOS binary packaging: push dependencies upstream (ARROW-16883)
- Windows binary packaging: push dependencies upstream (ARROW-16878)
- Update cloud/filesystem documentation (ARROW-16887)

### CODE REVIEW

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 1]
```diff
@@ -2690,15 +2690,16 @@ macro(resolve_dependency_absl)
...
+        strerror
         strings
         strings_internal
         symbolize
         synchronization
         throw_delegate
         time
-        time_zone)
+        time_zone
+        wyhash)
```
Comment (nealrichardson) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:2702]: Most of the cmake changes here seem to just be differences in sorting (due to my locale?), but the omission of `wyhash` seems to be legitimately new.

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 2]
```diff
@@ -2794,18 +2795,17 @@ macro(resolve_dependency_absl)
     endforeach()
 
     # Extracted the dependency information using the Abseil pkg-config files:
-    #   grep Requires $PREFIX/pkgconfig/absl_*.pc | \
+    #   grep Requires $PREFIX/lib/pkgconfig/absl_*.pc | \
     #   sed -e 's;.*/absl_;set_property(TARGET absl::;' \
     #       -e 's/.pc:Requires:/ PROPERTY INTERFACE_LINK_LIBRARIES /' \
-    #       -e 's/ = 20210324,//g' \
-    #       -e 's/ = 20210324//g' \
+    #       -E -e 's/ = 20[0-9]{6},?//g' \
```
Comment (nealrichardson) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:2801]: I simplified this regex so that it is robust to future abseil version bumps (at least until 2030!)

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 3]
```diff
@@ -3542,18 +3544,6 @@ macro(resolve_dependency_absl)
                    APPEND
                    PROPERTY INTERFACE_LINK_LIBRARIES ${CoreFoundation})
     endif()
-    set_property(TARGET absl::type_traits PROPERTY INTERFACE_LINK_LIBRARIES absl::config)
```
Comment (nealrichardson) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:3545]: These are duplicated above so I removed them here; I suspect these were an artifact of a bad merge

File: r/src/arrow_types.h [hunk 4]
```diff
@@ -69,6 +69,14 @@ namespace ds = ::arrow::dataset;
 namespace compute = ::arrow::compute;
 namespace fs = ::arrow::fs;
 
+#if defined(ARROW_R_WITH_GCS)
+namespace arrow {
+namespace fs {
+class GcsFileSystem;
```
Comment (nealrichardson) [inline r/src/arrow_types.h:75]: TODO(me): I think this should be put in arrow/filesystem/type_fwd.h

File: r/src/filesystem.cpp [hunk 5]
```diff
@@ -33,6 +33,12 @@ const char* r6_class_name<fs::FileSystem>::get(
     return "LocalFileSystem";
   } else if (type_name == "s3") {
     return "S3FileSystem";
+  } else if (type_name == "gcs") {
+    return "GcsFileSystem";
+  } else if (type_name == "abfs") {
+    return "AzureBlobFileSystem";
+  } else if (type_name == "hdfs") {
+    return "HadoopFileSystem";
```
Comment (nealrichardson) [inline r/src/filesystem.cpp:41]: Put them here for completeness but we should probably comment this block out for now since these R6 classes don't exist yet and the default fallback to FileSystem might allow things to Just Work if you're linking to a C++ library that has support for these built.

**Comment by nealrichardson** [comment]
R debian build is failing with an undefined symbol from absl, a different one from what I saw locally (on macOS): 

https://github.com/apache/arrow/runs/6989838112?check_suite_focus=true#step:5:2990

**Comment by nealrichardson** [comment]
@coryan any ideas about that undefined abseil symbol mentioned above? `undefined symbol: _ZN4absl12lts_2021110212RFC3339_fullE`

Similar issue in the homebrew build `Symbol not found: __ZN4absl12lts_2021110210FormatTimeENS0_11string_viewENS0_4TimeENS0_8TimeZoneE` See https://github.com/apache/arrow/pull/13407#issuecomment-1161923135

**Comment by coryan** [comment]
TL;DR; not much of an idea, sorry.

> any ideas about that undefined abseil symbol mentioned above?

I assume "above" refers to https://github.com/apache/arrow/runs/6989838112?check_suite_focus=true#step:5:2990 just let me know if I read that wrong.

> `undefined symbol: _ZN4absl12lts_2021110212RFC3339_fullE`

Seems like Abseil is not being linked?  The log says:

> -- All bundled static libraries: Snappy::snappy;thrift::thrift;jemalloc;mimalloc::mimalloc;lz4::lz4;zstd::libzstd;re2::re2;utf8proc::utf8proc;google-cloud-cpp::storage;google-cloud-cpp::common;aws-cpp-sdk-identity-management;aws-cpp-sdk-sts;aws-cpp-sdk-cognito-identity;aws-cpp-sdk-s3;aws-cpp-sdk-core;AWS::aws-c-event-stream;AWS::aws-checksums;AWS::aws-c-common

That seems to be missing Abseil, but I do not quite understand how the list of bundled dependencies is created.

> 
> Similar issue in the homebrew build `Symbol not found: __ZN4absl12lts_2021110210FormatTimeENS0_11string_viewENS0_4TimeENS0_8TimeZoneE` See [#13407 (comment)](https://github.com/apache/arrow/pull/13407#issuecomment-1161923135)

Probably a similar problem as above.  Less likely, it could be Abseil's sensitivity to C++11 vs. C++17 building (see abseil/abseil-cpp#696), I always think it could be that issue if `string_view`, `variant` or `any` are involved.

**Comment by nealrichardson** [comment]
@coryan thanks, that's helpful actually. I'll have to dig more on the bundled libraries thing. On the Homebrew build, abseil indeed is being built with C++17 by brew (https://github.com/Homebrew/homebrew-core/blob/master/Formula/abseil.rb#L31), and the arrow build is using C++11. What's odd is that we've been depending on it (transitively) for a while because Flight uses grpc (also built C++17 by brew), but I guess google-cloud-cpp hits different parts.

**Comment by nealrichardson** [comment]
Well, that didn't work, there must be some other piece I'm missing: 

```
-- Configuring done
CMake Error at cmake_modules/BuildUtils.cmake:112 (file):
  Error evaluating generator expression:
    $<TARGET_FILE:absl::memory>
  Target "absl::memory" is not an executable or library.
Call Stack (most recent call first):
  src/arrow/CMakeLists.txt:608 (create_merged_static_lib)
CMake Error at cmake_modules/BuildUtils.cmake:112 (file):
  Error evaluating generator expression:
    $<TARGET_FILE:absl::memory>
  Target "absl::memory" is not an executable or library.
Call Stack (most recent call first):
  src/arrow/CMakeLists.txt:608 (create_merged_static_lib)
```

**Comment by coryan** [comment]
I suspect `absl::memory` is a header-only library, so it does not have a `$<TARGET_FILE>`.

File: r/src/filesystem.cpp [hunk 6]
```diff
@@ -331,3 +337,77 @@ std::string fs___S3FileSystem__region(const std::shared_ptr<fs::S3FileSystem>& f
...
+
+  // /// \brief Default metadata for OpenOutputStream.
+  // ///
+  // /// This will be ignored if non-empty metadata is passed to OpenOutputStream.
+  if (!Rf_isNull(options["default_metadata"])) {
+    gcs_opts.default_metadata = strings_to_kvm(options["default_metadata"]);
+  }
+
+  auto io_context = arrow::io::IOContext(gc_memory_pool());
+  // TODO: S3FileSystem::Make returns a Result and uses ValueOrStop but this doesn't?
```
[the remark this answers is not in the document]
Comment (emkornfield) [reply r/src/filesystem.cpp:410]: I think this is because the GcsFileSystem does lazy connections so this can't currently error.  No objects to make them uniform though

[the remark this answers is not in the document]
Comment (nealrichardson) [reply r/src/filesystem.cpp:414]: @pitrou do you have an opinion on this?

[the remark this answers is not in the document]
Comment (pitrou) [reply r/src/filesystem.cpp:414]: Let's make it a `Result`, so that we can potentially raise errors on invalid parameters etc.

**Comment by coryan** [comment]
If you have a VM with abseil installed, you could do something like:

```sh
pkg-config --libs absl_memory absl_strings absl_str_format absl_time absl_variant absl_base absl_memory absl_optional absl_span absl_time absl_variant
-L/usr/local/lib64 -labsl_str_format_internal -labsl_bad_optional_access -labsl_time -labsl_civil_time -labsl_strings -labsl_strings_internal -lrt -labsl_base -labsl_spinlock_wait -labsl_int128 -labsl_throw_delegate -labsl_time_zone -labsl_bad_variant_access -labsl_raw_logging_internal -labsl_log_severity
```

Which (a) removes the header-only libraries, and (b) saves you some whack-a-moling (maybe).

**Comment by nealrichardson** [comment]
> If you have a VM with abseil installed, you could do something like:
> 
> ```shell
> pkg-config --libs absl_memory absl_strings absl_str_format absl_time absl_variant absl_base absl_memory absl_optional absl_span absl_time absl_variant
> -L/usr/local/lib64 -labsl_str_format_internal -labsl_bad_optional_access -labsl_time -labsl_civil_time -labsl_strings -labsl_strings_internal -lrt -labsl_base -labsl_spinlock_wait -labsl_int128 -labsl_throw_delegate -labsl_time_zone -labsl_bad_variant_access -labsl_raw_logging_internal -labsl_log_severity
> ```
> 
> Which (a) removes the header-only libraries, and (b) saves you some whack-a-moling (maybe).

Yeah I keep thinking I've reached the end (just one more library!) but that would be smarter. I have a local build of abseil already (from the arrow build), didn't think to point pkg-config at it. Thanks for the suggestion!

**Comment by nealrichardson** [comment]
@github-actions crossbow submit homebrew-r-autobrew

**Comment by nealrichardson** [comment]
For the R Windows packages, I need to use a static libcurl, and stack overflow tells me that I need to define `CURL_STATICLIB` for the google-cloud-cpp build. (The thing I just pushed won't work, it's not going to put it in the right place.) I don't see a way I can pass that in to the build macro from the outside (as in, from the `cmake` command that builds arrow), and I don't think we'd want to hard-code that. Is it possible in cmake to tell if `find_package(curl)` found a static or shared library? Then I think I could conditionally add that only when appropriate. @kou thoughts?

cc @jeroen in case you have different ideas

**Comment by wjones127** [comment]
(I created docs follow up: https://issues.apache.org/jira/browse/ARROW-16887)

File: r/configure.win [hunk 7]
```diff
@@ -104,6 +105,10 @@ function configure_dev() {
     PKG_CFLAGS="$PKG_CFLAGS -DARROW_R_WITH_S3"
   fi
 
+  if [ $(cmake_option ARROW_GCS) -eq 1 ]; then
+    PKG_CFLAGS="$PKG_CFLAGS -DARROW_R_WITH_GCS -DCURL_STATICLIB"
+  fi
```
Comment (wjones127) [inline r/configure.win:110]: I think we should be able to let pkg-config handle adding the flags and libraries for libcurl:

```suggestion
  if [ $(cmake_option ARROW_GCS) -eq 1 ]; then
    PKG_CFLAGS="$PKG_CFLAGS -DARROW_R_WITH_GCS"
    PKG_CONFIG_PACKAGES="$PKG_CONFIG_PACKAGES libcurl"
  fi
```

**Comment by kou** [comment]
We need to define `CURL_STATICLIB` on building google-cloud-cpp not Apache Arrow C++. But CMake's bundled `FindCURL.cmake` doesn't support `CURL_STATICLIB`: https://gitlab.kitware.com/cmake/cmake/-/issues/18378

How about the following?

```diff
diff --git a/cpp/cmake_modules/ThirdpartyToolchain.cmake b/cpp/cmake_modules/ThirdpartyToolchain.cmake
index 024859288b..299c131958 100644
--- a/cpp/cmake_modules/ThirdpartyToolchain.cmake
+++ b/cpp/cmake_modules/ThirdpartyToolchain.cmake
@@ -1067,6 +1067,15 @@ macro(find_curl)
                                                               "${CURL_LIBRARIES}")
     endif()
   endif()
+  if(WIN32 AND NOT CURL_STATIC_CHECKED)
+    get_target_property(CURL_LIBRARY CURL::libcurl IMPORTED_LOCATION)
+    get_filename_component(CURL_LIBRARY_EXT "${CURL_LIBRARY}" LAST_EXT)
+    if(CURL_LIBRARY_EXT STREQUAL "${CMAKE_STATIC_LIBRARY_SUFFIX}")
+      set_target_properties(CURL::libcurl
+                            PROPERTIES INTERFACE_COMPILE_DEFINITIONS "CURL_STATICLIB")
+    endif()
+    set(CURL_STATIC_CHECKED TRUE)
+  endif()
 endmacro()
 
 # ----------------------------------------------------------------------

```

File: r/configure.win (hunk 7 above)
Comment (wjones127) [reply r/configure.win:110]: Nevermind this might not help if libcurl is already statically linked in the arrow binaries.

**Comment by nealrichardson** [comment]
I turned on thirdparty build verbosity, and it looks like `-DCURL_STATICLIB` is only being added to the libarrow compilation, not the google_cloud_cpp compilation, so it still fails to link: https://github.com/apache/arrow/runs/7025275214?check_suite_focus=true#step:8:5397

@kou seems like this is close, any idea now?

**Comment by kou** [comment]
Oh, sorry. My patch was wrong. We need to do the similar thing in google-cloud-cpp no in our `ThirdpartyToolchain.cmake`. I'll push a fix to this branch later.

**Comment by kou** [comment]
I confirmed that we can solve the Curl's symbol problem by defining `CURL_STATICLIB` in google-cloud-cpp. I also opened a pull request to google-cloud-cpp: https://github.com/googleapis/google-cloud-cpp/pull/9345

R 4.2 CI job succeeded and R 4.1 CI job failed because R 4.2 uses ucrt and R 4.1 uses msvcrt. ucrt provides `_dupenv_s()` but msvcrt doesn't. google-cloud-cpp uses `_dupenv_s()`. Our existing CI jobs disable GCS because of this: https://github.com/apache/arrow/blob/master/.github/workflows/cpp.yml#L279-L281

Can we disable GCS for R 4.1?

**Comment by nealrichardson** [comment]
> I confirmed that we can solve the Curl's symbol problem by defining `CURL_STATICLIB` in google-cloud-cpp. I also opened a pull request to google-cloud-cpp: [googleapis/google-cloud-cpp#9345](https://github.com/googleapis/google-cloud-cpp/pull/9345)
> 

🎉 

> R 4.2 CI job succeeded and R 4.1 CI job failed because R 4.2 uses ucrt and R 4.1 uses msvcrt. ucrt provides `_dupenv_s()` but msvcrt doesn't. google-cloud-cpp uses `_dupenv_s()`. Our existing CI jobs disable GCS because of this: https://github.com/apache/arrow/blob/master/.github/workflows/cpp.yml#L279-L281
> 
> Error log: https://github.com/apache/arrow/runs/7036333396?check_suite_focus=true#step:11:589
> 
> ```
> C:/rtools40/mingw32/bin/../lib/gcc/i686-w64-mingw32/8.3.0/../../../../i686-w64-mingw32/bin/ld.exe: ../windows/arrow-8.0.0.9000/lib/i386/libarrow_bundled_dependencies.a(getenv.cc.obj):(.text+0x22): undefined reference to `_imp___dupenv_s'
> ```

Is this not something that can be patched/shimmed in?

> 
> Can we disable GCS for R 4.1?

Surely. I'll look for the cleanest way to do that.

**Comment by coryan** [comment]
> > I confirmed that we can solve the Curl's symbol problem by defining `CURL_STATICLIB` in google-cloud-cpp. I also opened a pull request to google-cloud-cpp: [googleapis/google-cloud-cpp#9345](https://github.com/googleapis/google-cloud-cpp/pull/9345)
> 
> 🎉

FWIW, that PR just got merged, and will be part of the next release, due early in 2022-07.

> 
> > R 4.2 CI job succeeded and R 4.1 CI job failed because R 4.2 uses ucrt and R 4.1 uses msvcrt. ucrt provides `_dupenv_s()` but msvcrt doesn't. google-cloud-cpp uses `_dupenv_s()`.

FYI: not anymore, that got fixed in in googleapis/google-cloud-cpp#9140 and is available in the latest release (v1.42.0).

**Comment by nealrichardson** [comment]
Cool, we should be able to pull https://github.com/coryan/google-cloud-cpp/commit/4b60f39b166855dc63bf85c382a5a47e23dd7817.diff into the patch we're applying already and solve both issues (for now, until we upgrade). I'm not sure how to construct the PATCH_COMMAND correctly to do that (my attempts are failing, I think), but it surely is possible.

**Comment by nealrichardson** [comment]
With the upgrade to google-cloud-cpp 1.42.0, the R Windows builds are now passing. 

Python failure is ARROW-16873

MingGW 32 failure is OOM https://github.com/apache/arrow/runs/7054579512?check_suite_focus=true#step:7:1019; maybe we need to turn off unity build there?

**Comment by nealrichardson** [comment]
@github-actions crossbow submit -g r

**Comment by kou** [comment]
> MinGW 64 failure is something about boost in compiling the GCS filesystem tests: https://github.com/apache/arrow/runs/7054579528?check_suite_focus=true#step:7:1310

We need a preparation to use boost/process.hpp with Mingw-w64 like https://github.com/apache/arrow/blob/master/cpp/src/arrow/filesystem/s3_test_util.cc#L24-L38 . We can handle it in a follow-up task.

**Comment by nealrichardson** [comment]
@github-actions crossbow submit test-r-offline-maximal test-r-depsource-system homebrew-r-brew

**Comment by nealrichardson** [comment]
> > MinGW 64 failure is something about boost in compiling the GCS filesystem tests: https://github.com/apache/arrow/runs/7054579528?check_suite_focus=true#step:7:1310
> 
> We need a preparation to use boost/process.hpp with Mingw-w64 like https://github.com/apache/arrow/blob/master/cpp/src/arrow/filesystem/s3_test_util.cc#L24-L38 . We can handle it in a follow-up task.

I made ARROW-16906 for that and added a TODO comment in the workflow.

**Comment by nealrichardson** [comment]
@github-actions crossbow submit homebrew-r-brew

**Comment by nealrichardson** [comment]
Ok @kou we've resolved all of the builds now, this is ready if you want to give it a final review/+1.

**Comment by nealrichardson** [comment]
One observation: the fully bundled Linux build (`R / rhub/debian-gcc-devel:latest`) seems to be 10-15 minutes slower with ARROW_GCS=ON. We may want to reconsider turning ARROW_GCS=ON without an explicit request to turn it on in the Linux source build (we can include it in the binaries), but that raises some other UX questions. cc @jonkeane

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 8]
```diff
@@ -4046,25 +4054,47 @@ macro(build_google_cloud_cpp_storage)
...
   add_library(google-cloud-cpp::storage STATIC IMPORTED)
   set_target_properties(google-cloud-cpp::storage
                         PROPERTIES IMPORTED_LOCATION
                                    "${GOOGLE_CLOUD_CPP_STATIC_LIBRARY_STORAGE}"
                                    INTERFACE_INCLUDE_DIRECTORIES
                                    "${GOOGLE_CLOUD_CPP_INCLUDE_DIR}")
+  # Update this from https://github.com/googleapis/google-cloud-cpp/blob/main/google/cloud/storage/google_cloud_cpp_storage.cmake
   set_property(TARGET google-cloud-cpp::storage
                PROPERTY INTERFACE_LINK_LIBRARIES
                         google-cloud-cpp::common
```
Comment (kou) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:4096]: We can remove this because `google-cloud-cpp-rest-internal` depends on `google-cloud-cpp::common`.

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 9]
```diff
@@ -4075,11 +4105,39 @@ macro(build_google_cloud_cpp_storage)
...
-                        OpenSSL::Crypto)
+                        OpenSSL::Crypto
+                        ZLIB::ZLIB)
   add_dependencies(google-cloud-cpp::storage google_cloud_cpp_ep)
 
-  list(APPEND ARROW_BUNDLED_STATIC_LIBS google-cloud-cpp::storage
+  list(APPEND
+       ARROW_BUNDLED_STATIC_LIBS
+       google-cloud-cpp::rest-internal
+       google-cloud-cpp::storage
```
Comment (kou) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:4115]: This may not be a problem but could you put `google-cloud-cpp::rest-internal` after `google-cloud-cpp::storage` because `google-cloud-cpp::storage` depends on `google-cloud-cpp::rest-internal`?

```suggestion
       google-cloud-cpp::storage
       google-cloud-cpp::rest-internal
```

File: r/configure [hunk 10]
```diff
@@ -229,44 +229,48 @@ if [ $? -eq 0 ]; then
   # Check for features
   LIB_DIR=`echo $PKG_DIRS | sed -e 's/^-L//'`
   ARROW_OPTS_CMAKE="$LIB_DIR/cmake/arrow/ArrowOptions.cmake"
-  # Check for Parquet
-  grep -i 'set(ARROW_PARQUET "ON")' $ARROW_OPTS_CMAKE >/dev/null 2>&1
-  if [ $? -eq 0 ]; then
+
+  arrow_built_with() {
+    # Function to check cmake options for features
+    grep -i 'set('"$1"' "ON")' $ARROW_OPTS_CMAKE >/dev/null 2>&1
```
Comment (kou) [inline r/configure:235]: FYI: We can use `grep -q ...` instead of `grep ... >/dev/null 2>&1`.

**Review by kou — APPROVED** [review]
+1

File: cpp/cmake_modules/ThirdpartyToolchain.cmake (hunk 8 above)
Comment (nealrichardson) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake:4096]: I guess so, but since it's copied from upstream, I'd rather leave it there for completeness unless it's harming anything.

File: r/configure (hunk 10 above)
Comment (nealrichardson) [reply r/configure:235]: `grep -q` doesn't suppress stderr though, and I want this to be completely silent
